src/index.ts

```diff
@@ -980,9 +1002,66 @@ class Envoy {
     }
   }
 
+
+  // --------------------------------------------------------------------------------------------
+
+  // Functions handling translation of dna_hash to hha_hash
+
+  async recordHha(hha_hash) {
+    // dna2hha is add-only
+    if (!this.hhaExists(hha_hash)) {
+      log.info("Retrieve the hosted app cell_data using the anonymous installed_app_id: '%s'", hha_hash);
+
+      const appInfo = await this.callConductor("app", { installed_app_id: hha_hash });
+
+      if (!appInfo) {
+        throw new Error(`No app found with installed_app_id: ${hha_hash}`);
+      }
+
+      // TODO but leave it for now: I am operating under the assumption that each dna_hash can be only in one app (identified by hha_hash)
+      // Does this need to change?
+      appInfo.cell_data.forEach(cell => {
+        let dna_hash_string = Codec.AgentId.encode(cell[0][0]); // cell[0][0] is binary buffer of dna_hash
+        this.dna2hha[dna_hash_string] = hha_hash;
+      });
+    }
+  }
+
+  hhaExists(hha_hash) {
+    return (Object.values(this.dna2hha).includes(hha_hash));
+  }
+
+  async signalHandler(signal) {
+    let cell_id = signal.data.cellId; // const signal: AppSignal = { type: msg.type , data: { cellId: [dna_hash, agent_id], payload: decodedPayload }};
+
+    // translate CellId->eventId
+    let event_id = this.cellId2eventId(cell_id);
+
+    log.debug(`Signal handler is emitting event ${event_id}`);
+    log.debug(`Signal content: ${signal.data.payload}`);
+    this.ws_server.emit(event_id, signal)
```

Comment:
Does this mean a hacked chaperone could subscribe to one of these event IDs and spy on signals directed at a different hosted user with the same host?

Reply:
Yes it does. But chaperone is running in a secure iframe that is controlled by Holo, so you cannot alter the code it is running..

BTW now that I think about it I am not sure if we have a check in envoy if the connection is in fact arriving from `https://chaprerone.holo.host` address. Will investigate.

Reply:
> But chaperone is running in a secure iframe that is controlled by Holo, so you cannot alter the code it is running..

You totally can. The iframe is controlled by the user's web browser. All you have to do is open inspect element and switch frames and you can run arbitrary code in the console as chaperone